File: common-java/src/main/java/com/common/kafka/MessageContextHolder.java
package com.common.kafka;

import java.util.HashMap;
import java.util.Map;

public class MessageContextHolder {
    private static final ThreadLocal<Map<String, String>> contextHolder = ThreadLocal.withInitial(HashMap::new);

    public static final String TXN_ID = "transactionId";
    public static final String MSG_ID = "messageId";
    public static final String SOURCE_ID = "sourceId";
    public static final String URI = "uri";

    public static void setTransactionId(String id) {
        contextHolder.get().put(TXN_ID, id);
    }

    public static String getTransactionId() {
        return contextHolder.get().get(TXN_ID);
    }

    public static void setMessageId(String id) {
        contextHolder.get().put(MSG_ID, id);
    }

    public static String getMessageId() {
        return contextHolder.get().get(MSG_ID);
    }

    public static void setSourceId(String id) {
        contextHolder.get().put(SOURCE_ID, id);
    }

    public static String getSourceId() {
        return contextHolder.get().get(SOURCE_ID);
    }

    public static void setUri(String uri) {
        contextHolder.get().put(URI, uri);
    }

    public static String getUri() {
        return contextHolder.get().get(URI);
    }

    public static void clear() {
        contextHolder.remove();
    }
}

File: common-js/MessageContextHolder.js
const { AsyncLocalStorage } = require('async_hooks');

const asyncLocalStorage = new AsyncLocalStorage();

class MessageContextHolder {
    static run(context, callback) {
        let store = asyncLocalStorage.getStore();
        if (!store) store = {};
        Object.assign(store, context);
        return asyncLocalStorage.run(store, callback);
    }

    static clear() {
        // In Node, Context automatically cleans up when the async execution tree resolves
    }

    static get() {
        return asyncLocalStorage.getStore() || {};
    }

    static getTransactionId() {
        return this.get().transactionId;
    }

    static setTransactionId(tid) {
        const store = this.get();
        if(store) store.transactionId = tid;
    }
    
    static getMessageId() { return this.get().messageId; }
    static setMessageId(id) { const store = this.get(); if(store) store.messageId = id; }

    static getSourceId() { return this.get().sourceId; }
    static setSourceId(id) { const store = this.get(); if(store) store.sourceId = id; }

    static getUri() { return this.get().uri; }
    static setUri(uri) { const store = this.get(); if(store) store.uri = uri; }
}

module.exports = MessageContextHolder;
